app/src/main/java/me/hufman/androidautoidrive/music/spotify/SpotifyWebApi.kt

```diff
@@ -0,0 +1,249 @@
+package me.hufman.androidautoidrive.music.spotify
+
+import android.app.NotificationChannel
+import android.app.NotificationManager
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import android.util.Log
+import androidx.core.app.NotificationCompat
+import com.adamratzman.spotify.*
+import com.adamratzman.spotify.models.Token
+import me.hufman.androidautoidrive.AppSettings
+import me.hufman.androidautoidrive.AppSettingsViewer
+import me.hufman.androidautoidrive.R
+import me.hufman.androidautoidrive.music.controllers.SpotifyAppController
+import me.hufman.androidautoidrive.phoneui.SpotifyAuthorizationActivity
+import net.openid.appauth.*
+
+/**
+ * Handles the logic around the Spotify Web API. This class is a singleton to prevent issues of having
+ * multiple [SpotifyClientAPI]s.
+ */
+class SpotifyWebApi private constructor(val context: Context) {
+	companion object {
+		const val TAG = "SpotifyWebApi"
+		const val NOTIFICATION_CHANNEL_ID = "SpotifyAuthorization"
+		const val NOTIFICATION_CHANNEL_NAME = "Spotify Authorization"
+		const val NOTIFICATION_REQ_ID = 56
+
+		private var webApiInstance: SpotifyWebApi? = null
+
+		/**
+		 * Retrieves the current [SpotifyWebApi] instance creating one if it there are no instances
+		 * currently running.
+		 */
+		fun getInstance(context: Context): SpotifyWebApi {
+			if (webApiInstance == null) {
+				webApiInstance = SpotifyWebApi(context)
+			}
+			return webApiInstance as SpotifyWebApi
+		}
+	}
+
+	// will be null if the user is not authenticated
+	private var webApi: SpotifyClientAPI? = null
+
+	private val authStateManager: SpotifyAuthStateManager
+	private val appSettingsViewer: AppSettingsViewer
+	private var getLikedSongsAttempted: Boolean = false
+	private var spotifyAppControllerCaller: SpotifyAppController? = null
+	var isUsingSpotify: Boolean = false
+
+	init {
+		Log.d(TAG, "Initializing for the first time")
+		authStateManager = SpotifyAuthStateManager(context)
+		appSettingsViewer = AppSettingsViewer()
+	}
+
+	/**
+	 * Sets the Liked Songs flag to false and sets the [SpotifyAppController] caller to null.
+	 */
+	fun clearGetLikedSongsAttemptedFlag() {
+		getLikedSongsAttempted = false
+		spotifyAppControllerCaller = null
+	}
+
+	fun getLikedSongs(spotifyAppController: SpotifyAppController): List<SpotifyMusicMetadata>?  {
```

Comment:
Since this function is called within an async scope in SpotifyAppController, can it return a kotlin coroutine Deferred object? Instead of remembering `getLikedSongsAttempted` and `spotifyAppControllerCaller`, it can then hold on to the Deferred object and complete it when the user reauthenticates. This will be a more extensible pattern to follow for future new APIs, like search (which is definitely a coroutine) and so on.
There could be a list of pending lambdas to call again when the API is reauthenticated, and these lambdas would hold on to their respective Deferred to complete.

Reply:
I've been playing around with this idea as I like the concept of storing the Deferred object only but there's a slight problem with this approach. The getLikedSongsAttempted() method is called from the SpotifyAppController as a part of the creation of the QueueMetadata so while we can store the Deferred object in the SpotifyWebApi and then call it when the Web API is reauthenticated (would be called in the initializeWebApi() method) we would just be calling the getLikedSongs() part which isn't responsible for building the QueueMetadata which would be out of the SpotifyAppController meaning that we'd still need the calling SpotifyAppController instance as this logic doesn't make sense to be handled in the SpotifyAppController. 

Additionally, this idea of calling failed API calls after reauthentication wasn't something I think we should have for anything other than something liked getLikedSongs() which is modifying the overall state of the music player. For instance if the user is trying to do something like search that requires the Web API and the Web API fails due to an authentication error and the user goes to do something else then reauthenticates later we don't want to do that search job since it is no longer relevant. I'd say it even goes for when the job still is relevant since the failure will most likely happen when the user is driving and then they won't be able to reauthenticate until they arrive at a destination or something. I think it's more appropriate to just fail and then the user will see that didn't work and after they reauthenticate they will try it again when it makes sense for them